### PR TITLE
basic styling + center

### PR DESCRIPTION
**Pull request by c-cube**
cc @grayswandyr please give it a try. It only works on the text part (line,lines,text)

2019-11-11-163938_822x247_scrot

### CODE REVIEW

**Comment by grayswandyr** [comment]
Hi, I will try tool into it this week, I will update you.

**Comment by c-cube** [comment]
preview of the "center" feature in this PR (along with bugfixes on master):

```
+---------------------------------------+
|the center of the triangle is        | |
|-------------------------------------+-|
|                                     | |
|                                     | |
|                                     | |
|                                     | |
|                                     | |
|                                     | |
|                                     |t|
|             lil' ol' me             |a|
|                                     |l|
|                                     |l|
|                                     | |
|                                     | |
|                                     | |
|                                     | |
|                                     | |
|                                     | |
|-------------------------------------+-|
|                    i'm aligned right| |
|-------------------------------------+-|
|loooooooooooooooooooooooooooooooooong| |
+---------------------------------------+

+----------+
|name1| foo|
|-----+----|
|    1|true|
+----------+
+-----------------------------+
|a |looooooooooooooooooooooooo|
|  |oonng                     |
|--+--------------------------|
|bx|                     +---+|
|  |                     |x|y||
|  |                     |-+-||
|  |                     |1|2||
|  |                     +---+|
+-----------------------------+
+------------------------------------+
|nice unicode! 💪                    |
|------------------------------------|
|+----------------------------------+|
||oï ωεird nums:|+-----------------+||
||π/2           ||+---------------+|||
||τ/4           |||+-------------+||||
||--------------||||sum=Σ_i a·xᵢ²|||||
||              ||||—————        |||||
||              ||||1+1          |||||
||  0           ||||-------------|||||
||  `+- 1       ||||           Ōₒ|||||
||   +- ω       ||||           À |||||
||      `+- ω²  |||+-------------+||||
||              ||+---------------+|||
||              |+-----------------+||
|+----------------------------------+|
+------------------------------------+


```

File: src/PrintBox_text.ml [hunk 1]
```diff
@@ -56,6 +97,7 @@ end = struct
     | Char of char
     | String of string
     | Str_slice of string * int * int
+    | Str_slice_bracket of string * string * int * int * string
```
Comment (Gbury) [inline src/PrintBox_text.ml]: Maybe use an inline record to distinguish all the integers and strings in there ?

File: src/PrintBox.ml [hunk 2]
```diff
@@ -1,16 +1,46 @@
...
   | Empty
-  | Text of string list
+  | Text of {
+      l: string list;
+      style: Style.t;
+    }
   | Frame of t
   | Pad of position * t (* vertical and horizontal padding *)
   | Align_right of t (* dynamic left-padding *)
+  | Center of t (* center vertically and horizontally *)
```
[the remark this answers is not in the document]
Comment (c-cube) [reply src/PrintBox.ml]: that's a good idea, we could have combinators for the 3 cases, too.

File: src/PrintBox.ml [hunk 3]
```diff
@@ -19,25 +49,30 @@ and t = view
 let empty = Empty
 let[@inline] view (t:t) : view = t
 
-let[@inline] line_ s = Text [s]
+let[@inline] line_ s = Text {l=[s]; style=Style.default}
 
-let line s =
+let line_with_style style s =
```
[the remark this answers is not in the document]
Comment (c-cube) [reply src/PrintBox.ml]: it slighty breaks compat, and more importantly, it's annoying to use in `|>` chains. Otherwise I'd have preferred it too! :)

**Comment by c-cube** [comment]
should be all good. @grayswandyr did you have time to take a look? :)

**Comment by grayswandyr** [comment]
I started but I'll only be able to to do it during this week-end. If it's not too long for you...

**Comment by c-cube** [comment]
That's ok, there's no hurry, and @Gbury is also reviewing the rendering code :)

**Comment by grayswandyr** [comment]
I have to adapt my code because I used to massage a string array before converting it into a `grid`. I need to refactor the code to be able to use the styling options (as a `t` is abstract, I can't apply styling options if the underlying string is equal to some thing).

**Comment by c-cube** [comment]
Note that there's now `grid_l` if you prefer lists :-)
The API is a bit richer now.

**Comment by grayswandyr** [comment]
Hi
from what I tested (in particular applying font styles and `init_grid`), everything was OK. Cool. Thanks!

**Pull request merged**